uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/BaseEditor.java

```diff
@@ -235,6 +246,60 @@ protected void makeMenuBar() {
         }
     }
 
+    protected Command getSaveAndRename() {
+
+        return getSaveAndRenameCommandBuilder()
+                .addPathSupplier(getPathSupplier())
+                .addValidator(getRenameValidator())
+                .addValidator(getSaveValidator())
+                .addRenameService(getSaveAndRenameServiceCaller())
+                .addMetadataSupplier(getMetadataSupplier())
+                .addContentSupplier(getContentSupplier())
+                .addIsDirtySupplier(isDirtySupplier())
+                .addSuccessCallback(onSuccess())
+                .build();
+    }
+
+    Supplier<Boolean> getSaveValidator() {
+
+        return () -> {
+
+            if (isReadOnly && versionRecordManager.isCurrentLatest()) {
+                baseView.alertReadOnly();
+                return false;
+            } else if (isReadOnly && !versionRecordManager.isCurrentLatest()) {
+                versionRecordManager.restoreToCurrentVersion();
+                return false;
+            }
+
+            if (concurrentUpdateSessionInfo != null) {
+                showConcurrentUpdatePopup();
+                return false;
+            } else {
+                return true;
+            }
+        };
+    }
+
+    protected ParameterizedCommand<Path> onSuccess() {
```

Comment:
It could be. Hash Codes are used by the (default) "is dirty" mechanism used when closing an editor to check if the content has changed. If an editor's underlying model does not implement ```hashCode()``` then I suspect if you make a change and close the editor the User is not prompted to save/discard changes. Any such issue however is out of the scope of this PR and new JIRAs for the applicable people raised.

Reply:
@manstis understood, will try to collect classes (through whole ensemble) where the hashCode is missing and thus are possibly dangerous for isDirty checking. Then file some jira possibly.